### PR TITLE
Verify assumed call requirements in CI builds

### PR DESCRIPTION
**Pull request by TheOneric**
See prior discussion in #820 and #822

Feel free to suggest other names for the macro

### CODE REVIEW

**Review by rcombs — COMMENTED** [review]
For reference, ffmpeg has `ASSERT_LEVEL`, where 0 enables only always-on asserts, 1 enables asserts that aren't in performance-critical code but aren't enabled by default, and 2 enables asserts in performance-critical code. We wanted something similar in the threading PR; maybe "convert ASSUMEs to asserts" could be ASSERT_LEVEL 3?

**Comment by TheOneric** [comment]
sounds good; adopted `ASSERT_LEVEL` and lacking a better place documented it in our readme

**Review by rcombs — APPROVED** [review]

**Comment by MrSmile** [comment]
Not sure about `ASSERT_LEVEL` without actual influence on `assert` (contrary to `NDEBUG`). I think it should be paired with corresponding `ass_assert` macro with level parameter (or `ass_assert_0`..`ass_assert_4`). Actually, if we're creating our own macro, we can merge `assert` and `ASSUME` into one.

**Comment by TheOneric** [comment]
> Not sure about `ASSERT_LEVEL` without actual influence on `assert` . I think it should be paired with corresponding `ass_assert` macro with level parameter (or `ass_assert_0`..`ass_assert_4`).

It affects assert insofar as it changes how many asserts are placed.
v1 of this patch just used `EXTRADEBUG`, but conditionally omitting some asserts came up before and in anticipation of future usage imho it makes sense to use a common knob for `assume → aassert` and assert omission.  
On the other hand, if we expect non-omission and assume-to-assert transforms to always appear together having 4 different levels would be needlessly complex and one singe `ifdef` condition work too. @rcombs how and why does FFmpeg use its three different assert levels?

Adding just the assume-to-assert capability now makes sense to me as we’re currently adding assumes and this can help catch bogus assumes early. I don’t want to couple this with reclassifying our existing asserts or adding new asserts for higher levels; this can happen later.

> Actually, if we're creating our own macro, we can merge assert and ASSUME into one.

In the default build configuration `ASSUME` and `ass_assert_{1,2,,3}` act differently enough to warrant a clear distinction imho

**Comment by rcombs** [comment]
ffmpeg's levels are:
- 0 is always on, and is used to catch conditions that should absolutely never be possible, where continuing would be unsafe, and crashing would be preferable even in a production build
- 1 is similar to a traditional C `assert`: it's expected to be on for most development builds, but off in most production builds; it's used for states that the program can probably safely recover from, but that probably indicate a problem that a developer would want to trap into a debugger to investigate
- 2 is further restricted to only trigger in builds where performance isn't a concern (usually this means e.g. `-g -O0` builds, not `-g -O3` builds that might be used for performance testing)

You could imagine tying these to different build configurations, like 0 for `release`, 1 for `reldebug`, and 2 for `debug`.

I think libass would benefit from having either this same set of 0-2, or perhaps 0-3 (where ASSUME is only converted to an assertion at level 3). Currently ffmpeg doesn't have an `ASSUME` equivalent (and perhaps it should).

**Comment by MrSmile** [comment]
> where ASSUME is only converted to an assertion at level 3

I think that assert should always convert to ASSUME when its level is disabled, separate macros do not make much sense. So usual assert = ass_assert_1, ASSUME = ass_assert_2 or 3.

**Comment by TheOneric** [comment]
I don’t mind much how the toggle and assert/assume macros are named; my primary concern is ensuring `ASSUME`s get some testing. I’m not sure what will be the best for planned, not yet existing optional asserts etc. If everyone agrees on one way to set it up now, that’ſ great and i can adopt this. But if the best way isn’t clear yet i suggest moving forward with just the `ASSUME` testing here and defer details of other assert levels to when they’ll actually get used.

**Comment by rcombs** [comment]
IMO we shouldn't automatically turn asserts into assumes, and if something's intended as an assume but also benefits from an assert in lower ASSERT_LEVELs, the developer should explicitly mark that. Some asserts may perform memory accesses or function calls that we don't want to ever happen in release builds, and the `ASSUME` may allow those accesses/calls to be emitted in some cases; it's also possible that we might have some asserts (at eg level 1 or 2) that we don't want to automatically become UB, but that we do want to trap into a debugger to diagnose on in development builds.

**Comment by MrSmile** [comment]
> But if the best way isn’t clear yet i suggest moving forward with just the `ASSUME` testing here and defer details of other assert levels to when they’ll actually get used.

In that case, I think it would be better to rename `ASSERT_LEVEL` to `DEBUG_LEVEL`.

> Some asserts may perform memory accesses or function calls that we don't want to ever happen in release builds

Fair point. But as I understand, both `__builtin_assume` and `__assume` do not actually evaluate expression, only `__builtin_unreachable` variant is a problem.

**Review by MrSmile — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -30,6 +30,17 @@ Windows. It is suited for static-only builds on any platform well supported by
...
+Macro defines
+-------------
+
+Unless developing libass there’s usually no need to manually define macros
+and there are no stability guarantees for these manual defines.
+
+- `DEBUG_LEVEL=0..3`
+  - `0` use the default set of asserts; implied if macro is not defined at all
+  - `1`, `2` *unused*
+  - `3` additionally assert assumptions usually pledged to the compiler for optimsation purposes
```
Comment (rcombs) [inline README.md:42]: typo: optimization

**Review by rcombs — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (TheOneric) [reply README.md]: fixed thx

**Pull request merged**